lib/client/keyagent.go

```diff
@@ -258,77 +259,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
+
+// checkHostCertificate validates a host certificate. First checks the
+// ~/.tsh/known_hosts cache and if not found, prompts the user to accept
+// or reject.
+func (a *LocalKeyAgent) checkHostCertificate(key ssh.PublicKey, addr string) bool {
+	// Check the local cache (where all Teleport CAs are placed upon login) to
+	// see if any of them match.
 	keys, err := a.keyStore.GetKnownHostKeys("")
 	if err != nil {
-		a.log.Error(err)
-		return trace.Wrap(err)
+		a.log.Errorf("Unable to fetch certificate authorities: %v.", err)
+		return false
 	}
-	a.log.Debugf("Got %d known hosts", len(keys))
 	for i := range keys {
-		if sshutils.KeysEqual(cert.SignatureKey, keys[i]) {
-			a.log.Debugf("Verified host %s", host)
-			return nil
+		if sshutils.KeysEqual(key, keys[i]) {
+			return true
 		}
 	}
-	// final step: if we have not seen the host key/cert before, lets ask the user if
-	// he trusts it, and add to the known_hosts if he says "yes"
-	if err = hostPromptFunc(host, key); err != nil {
-		// he said "no"
-		a.noHosts[host] = true
+
+	// If this certificate was not seen before, prompt the user essentially
+	// treating it like a key.
+	err = a.checkHostKey(addr, nil, key)
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+// checkHostCertificate validates a host key. First checks the
+// ~/.tsh/known_hosts cache and if not found, prompts the user to accept
+// or reject.
+func (a *LocalKeyAgent) checkHostKey(addr string, remote net.Addr, key ssh.PublicKey) error {
+	var err error
+
+	// Check if this exact host is in the local cache.
+	keys, _ := a.keyStore.GetKnownHostKeys(addr)
+	if len(keys) > 0 && sshutils.KeysEqual(key, keys[0]) {
+		a.log.Debugf("Verified host %s.", addr)
+		return nil
+	}
+
+	// If this key was not seen before, prompt the user with a fingerprint.
+	if a.hostPromptFunc != nil {
+		err = a.hostPromptFunc(addr, key)
+	} else {
+		err = a.defaultHostPromptFunc(addr, key)
+	}
+	if err != nil {
+		a.noHosts[addr] = true
 		return trace.Wrap(err)
 	}
-	// user said "yes"
-	err = a.keyStore.AddKnownHostKeys(host, []ssh.PublicKey{key})
+
+	// If the user trusts the key, store the key in the local known hosts
+	// cache ~/.tsh/known_hosts.
+	err = a.keyStore.AddKnownHostKeys(addr, []ssh.PublicKey{key})
 	if err != nil {
-		a.log.Warn(err)
+		a.log.Warnf("Failed to save the host key: %v.", err)
+		return trace.Wrap(err)
 	}
-	return err
+	return nil
+}
+
+// defaultHostPromptFunc is the default host key/certificates prompt.
+func (a *LocalKeyAgent) defaultHostPromptFunc(host string, key ssh.PublicKey) error {
+	userAnswer := "no"
+	if !a.noHosts[host] {
+		fmt.Printf("The authenticity of host '%s' can't be established. "+
+			"Its public key is:\n%s\nAre you sure you want to continue (yes/no)? ",
+			host, ssh.MarshalAuthorizedKey(key))
+
+		bytes := make([]byte, 12)
+		os.Stdin.Read(bytes)
+		userAnswer = strings.TrimSpace(strings.ToLower(string(bytes)))
+	}
+	ok, err := utils.ParseBool(userAnswer)
+	if err != nil || !ok {
```

Comment:
I think there is a mistake, `if err == nil and !ok` , you will return `trace.Wrap(nil)` what is the opposite of what we want?

Reply:
If I'm right, can you add the test for this to make sure we won't make this mistake again?